pkg/envoy/agent.go

```diff
@@ -17,22 +17,43 @@ package envoy
 import (
 	"context"
 	"errors"
+	"fmt"
+	"strconv"
+	"strings"
 	"time"
 
+	"istio.io/istio/pilot/pkg/util/sets"
+	"istio.io/istio/pkg/http"
 	"istio.io/pkg/log"
 )
 
 var errAbort = errors.New("epoch aborted")
 
 const errOutOfMemory = "signal: killed"
 
+// minDrainDuration is the minimum duration for which agent waits before it actively checks
+// for outstanding connections. After this period, agent terminates proxy whenever active
+// connections become zero.
+var minDrainDuration = 5 * time.Second
+
+var activeConnectionCheckDelay = 1 * time.Second
+
+var knownIstioListeners = sets.NewSet(
+	"listener.0.0.0.0_15006.downstream_cx_active", "listener.0.0.0.0_15021.downstream_cx_active",
+	"listener.0.0.0.0_15090.downstream_cx_active", "listener.admin.downstream_cx_active",
+	"listener.admin.main_thread.downstream_cx_active",
+)
+
 // NewAgent creates a new proxy agent for the proxy start-up and clean-up functions.
-func NewAgent(proxy Proxy, terminationDrainDuration time.Duration) *Agent {
+func NewAgent(proxy Proxy, terminationDrainDuration time.Duration, localhost string, adminPort uint16) *Agent {
 	return &Agent{
 		proxy:                    proxy,
 		statusCh:                 make(chan exitStatus, 1), // context might stop drainage
+		drainCh:                  make(chan bool),
```

Comment:
```suggestion
		drainCh:                  make(chan struct{}),
```

then use `close(drainCh)` to signal, not send